syntax = "proto3";
option go_package = "github.com/lexatic/web-backend/protos/lexatic-backend";
option java_package = "ai.rapida.sdk.artifacts";

import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";
message Criteria {
  string key = 1;
  string value = 2;
  string logic = 3;
}

message Error {
  uint64 errorCode = 1 [ jstype = JS_STRING ];
  string errorMessage = 2;
  string humanMessage = 3;
}

message Paginate {
  uint32 page = 1;
  uint32 pageSize = 2;
}

message Paginated {
  uint32 currentPage = 1;
  uint32 totalItem = 2;
}

message Ordering {
  string column = 1;
  string order = 2;
}

message User {
  uint64 id = 1 [ jstype = JS_STRING ];
  string name = 2;
  string email = 3;
  string role = 4;
  google.protobuf.Timestamp createdDate = 6;
  string status = 5;
}

message BaseResponse {
  int32 code = 1;
  bool success = 2;
  map<string, string> data = 3;
  Error error = 4;
}

message Metadata {
  uint64 id = 1 [ jstype = JS_STRING ];
  string key = 2;
  string value = 3;
}

message Variable {
  uint64 id = 1 [ jstype = JS_STRING ];
  string name = 2;
  string type = 3;
  optional string defaultValue = 4;
  optional string in = 5;
  bool required = 6;
  optional string label = 7;
}

message ProviderModelParameter {
  uint64 id = 1 [ jstype = JS_STRING ];
  uint64 providerModelVariableId = 2 [ jstype = JS_STRING ];
  string value = 3;
}
// message PromptVariable {
//   string name = 1;
//   string type = 2;
//   optional string defaultValue = 3;
//   string accepts = 4;
// }

message Provider {
  uint64 id = 1 [ jstype = JS_STRING ];
  string name = 2;
  string description = 3;
  string humanName = 4;
  string image = 5;
  string website = 6;
  string status = 7;
  repeated Variable connectConfiguration = 9;
}

message ProviderModelVariable {
  uint64 id = 1 [ jstype = JS_STRING ];
  uint64 providerModelId = 2 [ jstype = JS_STRING ];
  string key = 3;
  string name = 4;
  string description = 5;
  string defaultValue = 6;
  string type = 7;
  string place = 8;
  repeated Metadata metadatas = 9;
}

message ModelEndpoint { string endpoint = 1; }
message ProviderModel {
  uint64 id = 1 [ jstype = JS_STRING ];
  string name = 2;
  string description = 3;
  string humanName = 4;
  string category = 5;
  string status = 6;
  string owner = 7;
  Provider provider = 8;
  repeated ProviderModelVariable parameters = 9;
  repeated Metadata metadatas = 10;
  uint64 providerId = 11 [ jstype = JS_STRING ];
  repeated ModelEndpoint endpoints = 12;
}

message Tag {
  uint64 id = 1 [ jstype = JS_STRING ];
  repeated string tag = 2;
}

message Organization {
  uint64 id = 1 [ jstype = JS_STRING ];
  string name = 2;
  string description = 3;
  string industry = 4;
  string contact = 5;
  string size = 6;
}

message Metric {
  string name = 1;
  string value = 2;
  string description = 3;
}

message Content {
  string name = 1;
  // audio, image, video, text etc
  string contentType = 2;
  //  from raw string to url all can be
  string contentFormat = 3;
  //   actual content
  bytes content = 4;

  // added meta data incase you want to add something which is not supported in
  // future
  google.protobuf.Struct meta = 5;
}

message Message {
  string role = 1;
  repeated Content contents = 2;
}

message Knowledge {
  uint64 id = 1 [ jstype = JS_STRING ];
  string name = 2;
  string description = 3;
  string visibility = 4;
  string language = 5;
  uint64 embeddingProviderModelId = 6 [ jstype = JS_STRING ];
  ProviderModel embeddingProviderModel = 7;

  string status = 12;
  uint64 createdBy = 13 [ jstype = JS_STRING ];
  User createdUser = 14;
  uint64 updatedBy = 15 [ jstype = JS_STRING ];
  User updatedUser = 16;
  google.protobuf.Timestamp createdDate = 17;
  google.protobuf.Timestamp updatedDate = 18;

  //
  uint64 organizationId = 19 [ jstype = JS_STRING ];
  uint64 projectId = 20 [ jstype = JS_STRING ];
  Organization organization = 21;
  Tag knowledgeTag = 22;

  uint32 documentCount = 23;
  uint32 tokenCount = 24;
  uint32 wordCount = 25;

  uint64 embeddingProviderId = 26 [ jstype = JS_STRING ];
}

enum Source {
  WEB_PLUGIN = 0;
  RAPIDA_APP = 1;
  PYTHON_SDK = 2;
  NODE_SDK = 3;
}

enum Owner {
  RAPIDA = 0;
  UNKNOWN = 1;
  CLIENT = 2;
}

message ResourceIdentifier {
  string identifier = 1;
  Source source = 2;
  Owner owner = 3;
}